### PR TITLE
Add cloneTarStream,  pipePromise and multicastStream functions

### PR DESCRIPTION
**Pull request by pdcastro**
Change-type: minor
Signed-off-by: Paulo Castro <paulo@balena.io>

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -21,13 +41,323 @@ export function normalizeTarEntry(name: string): string {
...
+ * Return a promise that resolves when the operation "fully completes",
+ * i.e. when the 'finish' event has been emitted for every stream in the
+ * toStreams array.
+ * @param fromStream
+ * @param toStreams
+ */
+export function multicastStream(
+	fromStream: Readable,
+	toStreams: Writable[],
+): Bluebird<void> {
```
Comment (pdcastro) [inline lib/index.ts]: Conceptually, what `multicastStream` does is simple: it pipes a single source stream to a collection of destination streams provided by the caller, in small chunks and thus in a memory efficient way. It "listens" to the drain events of each destination stream separately, thus also works for mixed destination streams that consume the data at different rates. If any one destination applies backpressure, piping is paused for all destinations (to keep memory usage low). The implementation turned out to be more code than I expected, but hopefully not too hard to understand. I have written a few tests for it, including the case of destination streams that consume data at different rates, so I'm fairly confident it works. :-)
Sadly, the use case I wrote it for turned out not to be feasible. I had hoped to use it to replace some whole-file buffering in `resin-multibuild` ([here](https://github.com/balena-io-modules/resin-multibuild/blob/v1.0.0/lib/index.ts#L83-L90)), but then I learned (the hard way) that `tar-stream` treats `pack.entry(header, buf)` quite differently from `stream.pipe(pack.entry(header))`, enforcing stream backpressure only the to the latter while allowing the former to buffer as much memory as the system allows. Still, I think the `multicastStream` function is generic enough to have it as part of the toolkit (node-tar-utils).

File: lib/index.ts [hunk 2]
```diff
@@ -21,13 +41,323 @@ export function normalizeTarEntry(name: string): string {
...
+ * Create and return a promise that resolves to the pipeTo argument when the
+ * pipeFrom.pipe(pipeTo) operation completes.
+ * Warning: the destination stream (pipeTo) must be able to consume all the
+ * source data without stream backpressure blocking, or the returned promise
+ * will never resolve.
+ *
+ * @param pipeFrom Stream to pipe from
+ * @param pipeTo Stream to pipe to
+ * @param onFinish Optional 'finish' event handler for the pipeTo stream
+ * @param onError Optional 'error' event handler for the pipeTo stream
```
Comment (zvin) [inline lib/index.ts]: There's no `onFinish` or `onError` parameters below.

File: lib/index.ts (hunk 1 above)
Comment (zvin) [reply lib/index.ts]: You can pipe a readable stream to many writable streams just using `readable.pipe(writableX)` multiple times.

Comment (pdcastro) [reply lib/index.ts]: @zvin, I should stop adding you as reviewer to my pull requests! :-)
I was misled by this StackOverflow post where people seem to conclude that running `readable.pipe(writableX)` does not work: https://stackoverflow.com/questions/19553837/node-js-piping-the-same-readable-stream-into-multiple-writable-targets
But I suppose their circumstances are different. I have now pushed a much shorter implementation doing just `pipe` multiple times as suggested. My awesome tests have all passed, so I assume it just works. :-)

File: lib/index.ts (hunk 2 above)
Comment (pdcastro) [reply lib/index.ts]: 👍

**Pull request merged**